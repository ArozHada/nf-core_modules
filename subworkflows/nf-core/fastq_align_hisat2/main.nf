include { HISAT2_ALIGN            } from '../../../modules/nf-core/hisat2/align/main'
include { BAM_SORT_STATS_SAMTOOLS } from '../bam_sort_stats_samtools/main'

workflow FASTQ_ALIGN_HISAT2 {

    take:
    ch_reads       // channel: [ val(meta), path(reads) ]
    ch_index       // channel: [ path(index) ]      (/path/to/hisat2/index)
    ch_splicesites // channel: [ path(txt) ]        (/path/to/genome.splicesites.txt)
    ch_fasta    // channel: [ path(fasta) ]

    main:

    ch_versions = Channel.empty()


    //
    // Map reads with HISAT2
    //
    HISAT2_ALIGN ( ch_reads, ch_index, ch_splicesites )
    ch_versions = ch_versions.mix(HISAT2_ALIGN.out.versions.first())

    //
    // Sort, index BAM file and run samtools stats, flagstat and idxstats
    //
    BAM_SORT_STATS_SAMTOOLS ( HISAT2_ALIGN.out.bam, ch_fasta )
    ch_versions = ch_versions.mix(BAM_SORT_STATS_SAMTOOLS.out.versions)


    emit:
    orig_bam = HISAT2_ALIGN.out.bam                 // channel: [ val(meta), path(bam) ]
    summary  = HISAT2_ALIGN.out.summary             // channel: [ val(meta), path(log) ]
    fastq    = HISAT2_ALIGN.out.fastq               // channel: [ val(meta), path(fastq)]

    bam      = BAM_SORT_STATS_SAMTOOLS.out.bam      // channel: [ val(meta), path(bam) ]
    bai      = BAM_SORT_STATS_SAMTOOLS.out.bai      // channel: [ val(meta), path(bai) ]
    csi      = BAM_SORT_STATS_SAMTOOLS.out.csi      // channel: [ val(meta), path(csi) ]
    stats    = BAM_SORT_STATS_SAMTOOLS.out.stats    // channel: [ val(meta), path(stats) ]
    flagstat = BAM_SORT_STATS_SAMTOOLS.out.flagstat // channel: [ val(meta), path(flagstat) ]
    idxstats = BAM_SORT_STATS_SAMTOOLS.out.idxstats // channel: [ val(meta), path(idxstats) ]

    versions = ch_versions                          // channel: [ path(versions.yml) ]
}

