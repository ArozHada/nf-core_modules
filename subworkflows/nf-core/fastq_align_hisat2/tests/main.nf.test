nextflow_workflow {

    name "Test Workflow FASTQ_ALIGN_HISAT2"
    script "../main.nf"
    workflow "FASTQ_ALIGN_HISAT2"
    config "./nextflow.config"
    tag "subworkflows"
    tag "subworkflows_nfcore"
    tag "subworkflows/fastq_align_hisat2"
    tag "fastq_align_hisat2"
    tag "bam_sort_stats_samtools"
    tag "hisat2"
    tag "hisat2/align"
    tag "hisat2/build"
    tag "hisat2/extractsplicesites"

    setup {
        run("HISAT2_EXTRACTSPLICESITES") {
            script "../../../../modules/nf-core/hisat2/extractsplicesites/main.nf"
            process {
            """
            input[0] = [ [id:'test'],
                         file(params.test_data['sarscov2']['genome']['genome_gtf'], checkIfExists: true)
                       ]
            """
            }
        }

        run("HISAT2_BUILD") {
            script "../../../../modules/nf-core/hisat2/build/main.nf"
            process {
            """
            input[0] = [
                        [ id:'test' ],
                        [
                          file(params.test_data['sarscov2']['genome']['genome_fasta'], checkIfExists: true)
                        ]
                       ]
            input[1] = [ [id:'test'],
                         file(params.test_data['sarscov2']['genome']['genome_gtf'], checkIfExists: true)
                       ]
            input[2] = HISAT2_EXTRACTSPLICESITES.out.txt
            """
            }
        }
    }

    test("Single-End") {

        when {
            workflow {
                """
                input[0] = [
                            [ id:'test', single_end:true ], // meta map
                            [
                                file(params.test_data['sarscov2']['illumina']['test_1_fastq_gz'], checkIfExists: true)
                            ]
                           ]
                input[1] = HISAT2_BUILD.out.index
                input[2] = HISAT2_EXTRACTSPLICESITES.out.txt
                input[3] = [
                            [ id:'test' ],
                            [
                              file(params.test_data['sarscov2']['genome']['genome_fasta'], checkIfExists: true)
                            ]
                           ]
                """
            }
        }

        then {
            assertAll(
                { assert workflow.success },
                { assert snapshot(workflow.out.summary).match("se_summary") },
                { assert snapshot(workflow.out.fastq).match("se_fastq") },
                { assert snapshot(workflow.out.bai).match("se_bai") },
                { assert snapshot(workflow.out.csi).match("se_csi") },
                { assert snapshot(workflow.out.stats).match("se_stats") },
                { assert snapshot(workflow.out.flagstat).match("se_flagstat") },
                { assert snapshot(workflow.out.idxstats).match("se_idxstats") },
                { assert snapshot(workflow.out.versions).match("se_versions") },

                { assert workflow.out.orig_bam },
                { assert workflow.out.bam },
          )
        }
    }

    test("Paired-End") {

        when {
            workflow {
                """
                input[0] = [
                            [ id:'test', single_end:false ], // meta map
                            [
                              file(params.test_data['sarscov2']['illumina']['test_1_fastq_gz'], checkIfExists: true),
                              file(params.test_data['sarscov2']['illumina']['test_2_fastq_gz'], checkIfExists: true)
                            ]
                           ]
                input[1] = HISAT2_BUILD.out.index
                input[2] = HISAT2_EXTRACTSPLICESITES.out.txt
                input[3] = [
                            [ id:'test' ],
                            [
                              file(params.test_data['sarscov2']['genome']['genome_fasta'], checkIfExists: true)
                            ]
                           ]
                """
            }
        }

        then {
            assertAll(
                { assert workflow.success },
                { assert snapshot(workflow.out.summary).match("pe_summary") },
                { assert snapshot(workflow.out.fastq).match("pe_fastq") },
                { assert snapshot(workflow.out.bai).match("pe_bai") },
                { assert snapshot(workflow.out.csi).match("pe_csi") },
                { assert snapshot(workflow.out.stats).match("pe_stats") },
                { assert snapshot(workflow.out.flagstat).match("pe_flagstat") },
                { assert snapshot(workflow.out.idxstats).match("pe_idxstats") },
                { assert snapshot(workflow.out.versions).match("pe_versions") },

                { assert workflow.out.orig_bam },
                { assert workflow.out.bam },
          )
        }
    }
}
