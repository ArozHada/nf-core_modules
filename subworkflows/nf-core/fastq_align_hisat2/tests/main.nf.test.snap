{
    "se_versions": {
        "content": [
            [
                "versions.yml:md5,11b9cf11de1d9c428f6f2d1ff293c420",
                "versions.yml:md5,43bee7624dd94151feeac14346102228",
                "versions.yml:md5,493b15263efc8230741fc8c0a4ab378b",
                "versions.yml:md5,b4ccce0351e5718d36600858452dd4b1",
                "versions.yml:md5,b668538ba8fabc24e2d5ca5c0a1a8b87",
                "versions.yml:md5,e3a6719b98572eab19e8041c59031ac1"
            ]
        ],
        "timestamp": "2023-11-25T21:57:14.202284599"
    },
    "se_bai": {
        "content": [
            [
                [
                    {
                        "id": "test",
                        "single_end": true
                    },
                    "test.sorted.bam.bai:md5,af2f8016a65f96ba633f1d88ced0d919"
                ]
            ]
        ],
        "timestamp": "2023-11-25T21:57:14.125845434"
    },
    "pe_flagstat": {
        "content": [
            [
                [
                    {
                        "id": "test",
                        "single_end": false
                    },
                    "test.flagstat:md5,2fa0d90162a1b655863796c2a6bd8f45"
                ]
            ]
        ],
        "timestamp": "2023-11-25T21:57:40.923188208"
    },
    "se_summary": {
        "content": [
            [
                [
                    {
                        "id": "test",
                        "single_end": true
                    },
                    "test.hisat2.summary.log:md5,7b8a9e61b7646da1089b041333c41a87"
                ]
            ]
        ],
        "timestamp": "2023-11-25T21:57:14.112245095"
    },
    "pe_summary": {
        "content": [
            [
                [
                    {
                        "id": "test",
                        "single_end": false
                    },
                    "test.hisat2.summary.log:md5,9839b31db795958cc4b70711a3414e9c"
                ]
            ]
        ],
        "timestamp": "2023-11-25T21:57:40.884723151"
    },
    "pe_fastq": {
        "content": [
            [
                
            ]
        ],
        "timestamp": "2023-11-25T21:57:40.895166159"
    },
    "pe_idxstats": {
        "content": [
            [
                [
                    {
                        "id": "test",
                        "single_end": false
                    },
                    "test.idxstats:md5,1adb27b52d4d64b826f48b59d61dcd4d"
                ]
            ]
        ],
        "timestamp": "2023-11-25T21:57:40.932473447"
    },
    "se_fastq": {
        "content": [
            [
                
            ]
        ],
        "timestamp": "2023-11-25T21:57:14.122318774"
    },
    "se_stats": {
        "content": [
            [
                [
                    {
                        "id": "test",
                        "single_end": true
                    },
                    "test.stats:md5,de59b04575abceeb5f91d4088653baa9"
                ]
            ]
        ],
        "timestamp": "2023-11-25T21:57:14.133570664"
    },
    "se_csi": {
        "content": [
            [
                
            ]
        ],
        "timestamp": "2023-11-25T21:57:14.129887294"
    },
    "pe_csi": {
        "content": [
            [
                
            ]
        ],
        "timestamp": "2023-11-25T21:57:40.908427279"
    },
    "pe_stats": {
        "content": [
            [
                [
                    {
                        "id": "test",
                        "single_end": false
                    },
                    "test.stats:md5,0f64403e4f929933095d1c9c3c5ca2a3"
                ]
            ]
        ],
        "timestamp": "2023-11-25T21:57:40.914850298"
    },
    "se_flagstat": {
        "content": [
            [
                [
                    {
                        "id": "test",
                        "single_end": true
                    },
                    "test.flagstat:md5,6de3bfde9582ad2532033832091f5c46"
                ]
            ]
        ],
        "timestamp": "2023-11-25T21:57:14.138180623"
    },
    "se_idxstats": {
        "content": [
            [
                [
                    {
                        "id": "test",
                        "single_end": true
                    },
                    "test.idxstats:md5,2a5df85e0d90e55bb2b359f6e05d5fbb"
                ]
            ]
        ],
        "timestamp": "2023-11-25T21:57:14.146233523"
    },
    "pe_bai": {
        "content": [
            [
                [
                    {
                        "id": "test",
                        "single_end": false
                    },
                    "test.sorted.bam.bai:md5,ab51aa55d2da99b8be8e3ea13d112ecb"
                ]
            ]
        ],
        "timestamp": "2023-11-25T21:57:40.901188549"
    },
    "pe_versions": {
        "content": [
            [
                "versions.yml:md5,11b9cf11de1d9c428f6f2d1ff293c420",
                "versions.yml:md5,43bee7624dd94151feeac14346102228",
                "versions.yml:md5,493b15263efc8230741fc8c0a4ab378b",
                "versions.yml:md5,b4ccce0351e5718d36600858452dd4b1",
                "versions.yml:md5,b668538ba8fabc24e2d5ca5c0a1a8b87",
                "versions.yml:md5,e3a6719b98572eab19e8041c59031ac1"
            ]
        ],
        "timestamp": "2023-11-25T21:57:40.941526516"
    }
}